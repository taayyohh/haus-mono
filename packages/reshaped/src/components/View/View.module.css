.root {
  --rs-view-gaps-count-multiplier: 1;
  --rs-view-z: auto;
  --rs-view-gap-value: calc(var(--rs-view-gap) * var(--rs-unit-x1));

  z-index: var(--rs-view-z);
  gap: var(--rs-view-gap-value);
}

.item {
  --rs-view-item-gap: calc(
    var(--rs-unit-x1) * var(--rs-view-item-gap-before) - var(--rs-view-gap-value)
  );

  order: var(--rs-view-item-order);
}

.--padding {
  padding: calc(var(--rs-unit-x1) * var(--rs-view-p-vertical, var(--rs-view-p-top)))
    calc(var(--rs-unit-x1) * var(--rs-view-p-horizontal, var(--rs-view-p-start)));
}

.--padding-top {
  padding-top: calc(var(--rs-unit-x1) * var(--rs-view-p-top));
}

.--padding-bottom {
  padding-bottom: calc(var(--rs-unit-x1) * var(--rs-view-p-bottom));
}

.--padding-start {
  padding-inline-start: calc(var(--rs-unit-x1) * var(--rs-view-p-start));
}

.--padding-end {
  padding-inline-end: calc(var(--rs-unit-x1) * var(--rs-view-p-end));
}

.--animated {
  transition: var(--rs-duration-medium) var(--rs-easing-standard);
  transition-property: background-color, color, border-color;
}

@each $alignment in start, center, end {
  .--align-text-$alignment {
    text-align: $alignment;
  }
}

@each $color in neutral, positive, critical, primary {
  .--bg-$color {
    background-color: var(--rs-color-background-$color);
    color: var(--rs-color-on-background-$color);
  }
}

@each $color in neutral-faded, positive-faded, critical-faded, primary-faded, page,
  page-faded, disabled, disabled-faded, elevation-base, elevation-raised,
  elevation-overlay
{
  .--bg-$color {
    color: var(--rs-color-foreground-neutral);
    background-color: var(--rs-color-background-$color);
  }
}

.--bg-white {
  background-color: var(--rs-color-white);
  color: var(--rs-color-black);
}

.--bg-black {
  background-color: var(--rs-color-black);
  color: var(--rs-color-white);
}

.--shadow-raised {
  box-shadow: var(--rs-shadow-raised);
}

.--shadow-overlay {
  box-shadow: var(--rs-shadow-overlay);
}

.--bd {
  border: 1px solid var(--rs-view-border-color);
}

@each $color in neutral-faded, neutral, positive, positive-faded, critical, critical-faded,
  primary, primary-faded
{
  .--bd-$color {
    --rs-view-border-color: var(--rs-color-border-$color);
  }
}

.--bd-transparent {
  --rs-view-border-color: transparent;
}

.--overflow-hidden {
  overflow: hidden;
}

.--divided {
  --rs-view-gaps-count-multiplier: 2;
}

.--flex {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

/* Direction */
@responsive .--direction {
  @value column {
    flex-direction: column;
    flex-wrap: nowrap;

    & > .item--gap-before {
      margin-top: var(--rs-view-item-gap);
    }

    & > .item--gap-auto {
      margin-top: auto;
    }
  }

  @value column-reverse {
    flex-wrap: nowrap;
    flex-direction: column-reverse;

    & > .item--gap-before {
      margin-top: var(--rs-view-item-gap);
    }

    & > .item--gap-auto {
      margin-top: auto;
    }
  }

  @value row {
    flex-wrap: wrap;
    flex-direction: row;
    align-items: flex-start;

    & > .item--gap-before {
      margin-inline-start: var(--rs-view-item-gap);
    }

    & > .item--gap-auto {
      margin-inline-start: auto;
    }

    & > button {
      width: auto !important;
    }
  }

  @value row-reverse {
    flex-wrap: wrap;
    flex-direction: row-reverse;
    align-items: flex-start;

    & > .item--gap-before {
      margin-inline-start: var(--rs-view-item-gap);
    }

    & > .item--gap-auto {
      margin-inline-start: auto;
    }

    & > button {
      width: auto !important;
    }
  }
}

.--nowrap {
  flex-wrap: nowrap;

  & > * {
    flex-shrink: 0;
  }
}

.--wrap {
  flex-wrap: wrap;
}

@each $viewport in m, l, xl {
  @media (--rs-viewport-$viewport) {
    .--nowrap-false--$(viewport),
    .--wrap-true--$(viewport) {
      flex-wrap: wrap;
    }

    .--nowrap-true--$(viewport),
    .--wrap-false--$(viewport) {
      flex-wrap: nowrap;
    }
  }
}

/* Alignment */
@responsive .--align {
  @value start {
    align-items: flex-start !important;
  }

  @value end {
    align-items: flex-end !important;
  }

  @value center {
    align-items: center !important;
  }

  @value stretch {
    align-items: stretch !important;
  }

  @value baseline {
    align-items: baseline !important;
  }
}

@responsive .--justify {
  @value start {
    justify-content: flex-start;
  }

  @value end {
    justify-content: flex-end;
  }

  @value center {
    justify-content: center;
  }
}

.divider {
  align-self: stretch;
}

@responsive .item--grow {
  @value true {
    flex-grow: 1;

    /* Fixes nested View with nowrap inside the grow item */
    min-width: 0;
    flex-shrink: 1;
    flex-basis: 0;
  }

  @value false {
    flex-grow: 0;
    flex-shrink: 0;
    flex-basis: auto;
    min-width: auto;
  }
}

@each $columns in 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 {
  .item--columns-$columns {
    --rs-view-columns-count: calc(12 / $columns);
    --rs-view-gaps-count: calc(
      (var(--rs-view-columns-count) - 1) * var(--rs-view-gaps-count-multiplier)
    );

    width: calc(
      (100% - var(--rs-view-gap-value) * var(--rs-view-gaps-count)) /
        var(--rs-view-columns-count) -
        var(--rs-view-item-gap, 0px)
    );
  }
}

.item--columns-auto {
  width: auto;
}

@each $viewport in m, l, xl {
  @media (--rs-viewport-$viewport) {
    @each $columns in 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 {
      .item--columns-$(columns)--$(viewport) {
        --rs-view-columns-count: calc(12 / $columns);
        --rs-view-gaps-count: calc(
          (var(--rs-view-columns-count) - 1) * var(--rs-view-gaps-count-multiplier)
        );

        width: calc(
          (100% - var(--rs-view-gap-value) * var(--rs-view-gaps-count)) /
            var(--rs-view-columns-count) -
            var(--rs-view-item-gap, 0px)
        );
      }

      .item--columns-auto--$(viewport) {
        width: auto;
      }
    }
  }
}

/* Variables */
@responsive .--padding {
  @variable --rs-view-p-vertical 0;
  @variable --rs-view-p-horizontal 0;
}

@responsive .--padding-top {
  @variable --rs-view-p-top 0;
}

@responsive .--padding-bottom {
  @variable --rs-view-p-bottom 0;
}

@responsive .--padding-start {
  @variable --rs-view-p-start 0;
}

@responsive .--padding-end {
  @variable --rs-view-p-end 0;
}

@responsive .root {
  @variable --rs-view-gap 0;
}

@responsive .root > .item {
  @variable --rs-view-item-order 0;
  @variable --rs-view-item-gap-before initial;
}
