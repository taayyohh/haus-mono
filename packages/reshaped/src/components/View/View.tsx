import React from 'react'
import { classNames, responsiveClassNames, responsiveVariables } from 'utilities/helpers'
import Divider, { DividerProps } from 'components/Divider'
import Hidden from 'components/Hidden'
import type * as G from 'types/global'
import type * as T from './View.types'
import s from './View.module.css'
import getRadiusStyles from 'styles/radius'
import getBleedStyles from 'styles/bleed'
import getWidthStyles from 'styles/width'
import getHeightStyles from 'styles/height'
import getMaxWidthStyles from 'styles/maxWidth'
import getMaxHeightStyles from 'styles/maxHeight'
import getPositionStyles from 'styles/position'
import getInsetStyles from 'styles/inset'
import getAspectRatioStyles from 'styles/aspectRatio'

const ViewItem = <As extends keyof JSX.IntrinsicElements = 'div'>(
  props: T.ItemProps<As>
) => {
  const {
    columns,
    grow,
    gapBefore,
    as: TagName = 'div' as any,
    order,
    children,
    className,
    attributes,
  } = props
  const itemClassNames = classNames(
    s.item,
    className,
    gapBefore === 'auto' && s['item--gap-auto'],
    gapBefore !== undefined && s['item--gap-before'],
    ...responsiveClassNames(s, 'item--grow', grow),
    ...responsiveClassNames(s, 'item--columns', columns)
  )

  const itemVariables = {
    ...responsiveVariables('--rs-view-item-order', order),
    ...responsiveVariables('--rs-view-item-gap-before', gapBefore),
  }

  return (
    <TagName
      {...attributes}
      style={{ ...attributes?.style, ...itemVariables }}
      className={itemClassNames}
    >
      {children}
    </TagName>
  )
}

const View = <As extends keyof JSX.IntrinsicElements = 'div'>(props: T.Props<As>) => {
  const {
    /**
     * Layout props
     */
    align,
    justify,
    wrap,
    gap,
    height,
    width,
    aspectRatio,
    maxHeight,
    maxWidth,
    padding,
    paddingInline,
    paddingBlock,
    paddingBottom,
    paddingEnd,
    paddingStart,
    paddingTop,
    bleed,

    /**
     * Style props
     * */
    animated,
    backgroundColor,
    borderColor,
    borderRadius,
    shadow,
    textAlign,
    overflow,
    position = 'relative',
    inset,
    insetTop,
    insetBottom,
    insetStart,
    insetEnd,
    zIndex,

    /**
     * Item prop
     */
    grow,

    /**
     * Using any here to let TS save on type resolving, otherwise TS throws an error due to the type complexity
     * It still resolves the attributes correctly based on the tag
     */
    as: TagName = 'div' as any,
    children,
    divided,
    className,
    attributes,
  } = props
  const isFlex = !!align || !!justify || !!gap || !!props.direction
  const direction = props.direction || (isFlex ? 'column' : undefined)
  const radiusStyles = getRadiusStyles(borderRadius)
  const bleedStyles = getBleedStyles(bleed)
  const widthStyles = getWidthStyles(width)
  const heightStyles = getHeightStyles(height)
  const maxWidthStyles = getMaxWidthStyles(maxWidth)
  const maxHeightStyles = getMaxHeightStyles(maxHeight)
  const positionStyles = getPositionStyles(position)
  const insetStyles = getInsetStyles(inset)
  const insetTopStyles = getInsetStyles(insetTop, 'top')
  const insetBottomStyles = getInsetStyles(insetBottom, 'bottom')
  const insetStartStyles = getInsetStyles(insetStart, 'start')
  const insetEndStyles = getInsetStyles(insetEnd, 'end')
  const aspectRatioStyles = getAspectRatioStyles(aspectRatio)

  let renderedItemIndex = 0
  // If wrap is not defined, it can be set based on item grow and split usage
  let nowrap

  const renderDivider: T.RenderDivider = ({ className, key }) => {
    const dividerClassName = classNames(s.divider, className)
    let isDividerVertical: DividerProps['vertical'] = false

    if (typeof direction === 'string' && direction.startsWith('row')) {
      isDividerVertical = true
    } else if (direction) {
      const viewports = Object.keys(direction) as Array<keyof G.ResponsiveOnly<string>>
      isDividerVertical = viewports.reduce((acc, viewport) => {
        const viewportDirection = (direction as G.ResponsiveOnly<T.Direction>)[viewport]

        if (!viewportDirection) return acc

        return {
          ...acc,
          [viewport]: viewportDirection.startsWith('row'),
        }
      }, {} as G.ResponsiveOnly<boolean>)
    }

    return (
      <div className={dividerClassName} key={`${key}-divider`}>
        <Divider vertical={isDividerVertical} blank />
      </div>
    )
  }

  const renderItem: T.RenderItem = ({ className, child, index }) => {
    const isItem = child.type === ViewItem
    const key = child.key || index
    const dividerElement = !!index && divided && renderDivider({ className, key })
    let itemElement

    if (isItem) {
      itemElement = React.cloneElement(child, {
        className: classNames(className, child.props.className),
      })
    } else if (className || !React.isValidElement(child)) {
      itemElement = (
        <div className={className} key={key}>
          {child}
        </div>
      )
    } else {
      itemElement = child
    }

    // Passing grow here because it's responsive and nowrap should follow it
    if (isItem && child.props?.grow) nowrap = child.props.grow
    if (isItem && child.props?.gap === 'auto') nowrap = true

    return [dividerElement, itemElement]
  }

  const formattedChildren = React.Children.map(children, (child: any, index) => {
    if (!child) return null

    // Ignore the indices of the items that rendered nothing
    const renderedIndex = renderedItemIndex
    renderedItemIndex += 1

    if (child.type === Hidden && typeof child.props.children !== 'function') {
      const { children: hiddenChild, ...hiddenProps } = child.props
      const key = child.key || index

      return (
        <Hidden {...hiddenProps} key={key}>
          {(className) =>
            renderItem({ className, child: hiddenChild, index: renderedIndex })
          }
        </Hidden>
      )
    }

    return renderItem({ child, index: renderedIndex })
  })

  // Classnames and attributes are written here so we can assign nowrap to the root element based on the children
  const rootClassNames = classNames(
    s.root,
    className,
    radiusStyles?.classNames,
    positionStyles?.classNames,
    bleedStyles?.classNames,
    widthStyles?.classNames,
    heightStyles?.classNames,
    aspectRatioStyles?.classNames,
    maxWidthStyles?.classNames,
    maxHeightStyles?.classNames,
    insetStyles?.classNames,
    insetTopStyles?.classNames,
    insetBottomStyles?.classNames,
    insetStartStyles?.classNames,
    insetEndStyles?.classNames,
    textAlign && s[`--align-text-${textAlign}`],
    backgroundColor && s[`--bg-${backgroundColor}`],
    borderColor && s[`--bd-${borderColor}`],
    borderColor && s['--bd'],
    shadow && s[`--shadow-${shadow}`],
    overflow && s[`--overflow-${overflow}`],
    animated && s['--animated'],
    divided && s['--divided'],
    (padding !== undefined ||
      paddingInline !== undefined ||
      paddingBlock !== undefined) &&
      s['--padding'],
    paddingBottom !== undefined && s['--padding-bottom'],
    paddingEnd !== undefined && s['--padding-end'],
    paddingStart !== undefined && s['--padding-start'],
    paddingTop !== undefined && s['--padding-top'],
    (isFlex || nowrap) && s['--flex'],
    ...responsiveClassNames(s, '--direction', direction),
    ...responsiveClassNames(s, '--align', align),
    ...responsiveClassNames(s, '--justify', justify),
    // Wrap and nowrap are separate here because inverting any of them could result into a false value which will be ignored by classNames
    ...responsiveClassNames(s, '--nowrap', nowrap || wrap === false),
    ...responsiveClassNames(s, '--wrap', wrap),
    // Item classnames
    ...responsiveClassNames(s, 'item--grow', grow)
  )

  const rootVariables = {
    ...attributes?.style,
    ...responsiveVariables('--rs-view-gap', gap),
    ...responsiveVariables('--rs-view-p-vertical', paddingBlock || padding),
    ...responsiveVariables('--rs-view-p-horizontal', paddingInline || padding),
    ...responsiveVariables('--rs-view-p-bottom', paddingBottom),
    ...responsiveVariables('--rs-view-p-top', paddingTop),
    ...responsiveVariables('--rs-view-p-start', paddingStart),
    ...responsiveVariables('--rs-view-p-end', paddingEnd),
    ...bleedStyles?.variables,
    ...widthStyles?.variables,
    ...heightStyles?.variables,
    ...aspectRatioStyles?.variables,
    ...maxWidthStyles?.variables,
    ...maxHeightStyles?.variables,
    ...insetStyles?.variables,
    ...insetTopStyles?.variables,
    ...insetBottomStyles?.variables,
    ...insetStartStyles?.variables,
    ...insetEndStyles?.variables,
    ...(zIndex ? { '--rs-view-z': zIndex } : {}),
  }

  return (
    <TagName {...attributes} className={rootClassNames} style={rootVariables}>
      {formattedChildren}
    </TagName>
  )
}

View.Item = ViewItem
export default View
