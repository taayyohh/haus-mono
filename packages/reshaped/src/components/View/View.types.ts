import React from "react";
import type * as G from "types/global";
import type * as TStyles from "styles/types";

type Columns = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | "auto";
export type Direction = "row" | "column" | "row-reverse" | "column-reverse";

export type Props<TagName extends keyof JSX.IntrinsicElements = "div"> = {
	children?: React.ReactNode;
	as?: TagName;
	divided?: boolean;
	direction?: G.Responsive<Direction>;
	gap?: G.Responsive<number>;
	wrap?: G.Responsive<boolean>;
	align?: G.Responsive<"center" | "start" | "end" | "stretch" | "baseline">;
	justify?: G.Responsive<"center" | "start" | "end">;
	height?: G.Responsive<string | number>;
	width?: G.Responsive<string | number>;
	aspectRatio?: G.Responsive<number>;
	maxHeight?: G.Responsive<string | number>;
	maxWidth?: G.Responsive<string | number>;
	padding?: G.Responsive<number>;
	paddingTop?: G.Responsive<number>;
	paddingBottom?: G.Responsive<number>;
	paddingStart?: G.Responsive<number>;
	paddingEnd?: G.Responsive<number>;
	paddingInline?: G.Responsive<number>;
	paddingBlock?: G.Responsive<number>;
	bleed?: G.Responsive<number>;
	textAlign?: "center" | "start" | "end";
	backgroundColor?:
		| "neutral"
		| "neutral-faded"
		| "critical"
		| "critical-faded"
		| "positive"
		| "positive-faded"
		| "primary"
		| "primary-faded"
		| "elevation-base"
		| "elevation-raised"
		| "elevation-overlay"
		| "page"
		| "page-faded"
		| "disabled"
		| "disabled-faded"
		| "white"
		| "black";
	borderColor?:
		| "neutral"
		| "neutral-faded"
		| "critical"
		| "critical-faded"
		| "positive"
		| "positive-faded"
		| "primary"
		| "primary-faded"
		| "transparent";
	borderRadius?: G.Responsive<TStyles.Radius>;
	position?: G.Responsive<TStyles.Position>;
	inset?: G.Responsive<number>;
	insetStart?: G.Responsive<number>;
	insetEnd?: G.Responsive<number>;
	insetTop?: G.Responsive<number>;
	insetBottom?: G.Responsive<number>;
	zIndex?: number;
	shadow?: "raised" | "overlay";
	overflow?: "hidden";
	animated?: boolean;
	className?: G.ClassName;
	attributes?: G.Attributes<TagName>;
} & Pick<ItemProps, "grow">;

export type ItemProps<TagName extends keyof JSX.IntrinsicElements = "div"> = {
	order?: G.Responsive<number>;
	columns?: G.Responsive<Columns>;
	grow?: G.Responsive<boolean>;
	gapBefore?: G.Responsive<number> | "auto";
	as?: TagName;
	attributes?: G.Attributes<TagName>;
	className?: G.ClassName;
	children?: React.ReactNode;
};

export type RenderItem = (args: {
	className?: string;
	child: any;
	index: number;
}) => React.ReactNode;

export type RenderDivider = (args: { className?: string; key: string }) => React.ReactNode;
