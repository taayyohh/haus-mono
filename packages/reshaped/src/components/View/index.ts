export { default } from "./View";
export type { Props as ViewProps, ItemProps as ViewItemProps } from "./View.types";
