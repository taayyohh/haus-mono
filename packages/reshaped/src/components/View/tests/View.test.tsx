import React from "react";
import { render, screen } from "@testing-library/react";
import View from "components/View";

const fixtures = {
	content1: "Content 1",
	content2: "Content 2",
	className: "test-className",
	id: "test-id",
};

describe("Utilities/View", () => {
	test("renders children", () => {
		render(
			<View as="ul">
				<div>{fixtures.content1}</div>
				<View.Item as="li">{fixtures.content2}</View.Item>
			</View>
		);

		expect(screen.getByText(fixtures.content1)).toBeInTheDocument();
		expect(screen.getByText(fixtures.content2)).toBeInTheDocument();
		expect(screen.getByRole("list")).toBeInTheDocument();
		expect(screen.getAllByRole("listitem").length).toEqual(1);
	});

	test("applies className and attributes", () => {
		const { container } = render(
			<View
				className={fixtures.className}
				attributes={{ id: fixtures.id, style: { padding: 20 } }}
			/>
		);

		expect(container.firstChild).toHaveClass(fixtures.className);
		expect(container.firstChild).toHaveAttribute("id", fixtures.id);
	});

	test("applies className and attributes to View.Item", () => {
		const { container } = render(
			<View.Item
				className={fixtures.className}
				attributes={{ id: fixtures.id, style: { padding: 20 } }}
			/>
		);

		expect(container.firstChild).toHaveClass(fixtures.className);
		expect(container.firstChild).toHaveAttribute("id", fixtures.id);
		expect(container.firstElementChild?.getAttribute("style")).toMatchSnapshot();
	});
});
