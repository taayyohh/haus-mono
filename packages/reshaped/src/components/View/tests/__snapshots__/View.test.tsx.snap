// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Utilities/View > applies className and attributes to View.Item 1`] = `"padding: 20px;"`;
